File: src/graphics/GravityVectorsEffect.java
package graphics;

import java.awt.Color;
import java.awt.Graphics;
import java.util.List;

import structures.Ball;
import structures.Body;
import structures.CalcHelp;
import structures.Level;
import structures.Moon;
import structures.Point2d;

/**
 * 
 * @author Sean Lewis
 *
 */
public class GravityVectorsEffect extends GraphicEffect {

	public void draw(Level level, Graphics g) {
		
		List<Body> bodies = level.getBodies();
		Ball ball = level.getBall();
		double screenXShift = level.getScreenXShift();
		double screenYShift = level.getScreenYShift();

		g.setColor(Color.white);

		Point2d ballCent = new Point2d(ball.getCenter().x, ball.getCenter().y);
		for (Body b : bodies) {
			Point2d bodyCent = b.getCenter();
			double angle = CalcHelp.getAngle(ballCent, bodyCent);
			double length = level.getGravityStrength() * ArrowLength
					* b.getRadius() / ballCent.getDistance(bodyCent) + 5;
			Point2d p2 = new Point2d(ballCent.x + screenXShift + length
					* Math.cos(angle), ballCent.y + length * -Math.sin(angle)
					+ screenYShift);
			Point2d ballPt = new Point2d(ballCent.x + screenXShift, ballCent.y
					+ screenYShift);
			drawArrow(ballPt, p2, ArrowDistanceFromBall, ArrowSize, g);
			
			for (Moon m : b.getMoons()) {
				Point2d moonCent = m.getCenter();
				angle = CalcHelp.getAngle(ballCent, moonCent);
				length = level.getGravityStrength() * ArrowLength
						* m.getRadius() / ballCent.getDistance(moonCent) + 5;
				p2 = new Point2d(ballCent.x + screenXShift + length
						* Math.cos(angle), ballCent.y + length
						* -Math.sin(angle) + screenYShift);
				drawArrow(ballPt, p2, ArrowDistanceFromBall, ArrowSize, g);
			}
			
		}

	}

}

File: src/graphics/GravityVectorsEffect.scala
package graphics

import java.awt.Color
import java.awt.Graphics
import java.util.List
import structures._

/**
 * Holder class for gravitational vectors effect.
 * @author Sean Lewis
 */
object GravityVectorsEffect {

  /**
   * Draws the vectors from the ball.
   * @param level the current Level in the game
   * @param g the Graphics component to draw with
   */
  def draw(level: Level, g: Graphics) = {
    g.setColor(Color.white);
    val ball = level.getBall()
    val ballCent = new Point2d(ball.getCenter.x, ball.getCenter.y)
    val bodies = level.getBodies()
    val shift = level.getShift()
    val ballPt = ballCent + shift

    def drawVector(b: Body) = {
      val angle = CalcHelp.getAngle(ballCent, b.getCenter());
      val length = 5 + level.getGravityStrength() * ArrowLength *
        b.getRadius() / ballCent.distance(b.getCenter)
      val dx = shift.x + length *  math.cos(angle)
      val dy = shift.y + length * -math.sin(angle)
      val p2 = ballCent.translate(dx, dy)
      drawArrow(ballPt, p2, 4, 12, g)
    }

    for (i <- 0 until bodies.size()) {
      val b = bodies.get(i)
      drawVector(b)
      for (i <- 0 until b.getMoons().size()) {
        val m = b.getMoons().get(i)
        drawVector(m)
      }
    }

  }

}